homeassistant/components/notify/__init__.py

```diff
@@ -58,88 +55,160 @@
+    if (
+        NOTIFY_SERVICES not in hass.data
+        or integration_name not in hass.data[NOTIFY_SERVICES]
+    ):
+        return False
+
+    return True
 
 
-async def _async_notify_message_service(hass, service, notify_service, targets):
-    """Handle sending notification message service calls."""
-    kwargs = {}
-    message = service.data[ATTR_MESSAGE]
-    title = service.data.get(ATTR_TITLE)
+class BaseNotificationService:
+    """An abstract class for notification services."""
 
-    if title:
-        title.hass = hass
-        kwargs[ATTR_TITLE] = title.async_render()
+    hass: Optional[HomeAssistantType] = None
+
+    def send_message(self, message, **kwargs):
+        """Send a message.
 
-    if targets.get(service.service) is not None:
-        kwargs[ATTR_TARGET] = [targets[service.service]]
-    elif service.data.get(ATTR_TARGET) is not None:
-        kwargs[ATTR_TARGET] = service.data.get(ATTR_TARGET)
+        kwargs can contain ATTR_TITLE to specify a title.
+        """
+        raise NotImplementedError()
 
-    message.hass = hass
-    kwargs[ATTR_MESSAGE] = message.async_render()
-    kwargs[ATTR_DATA] = service.data.get(ATTR_DATA)
+    async def async_send_message(self, message: Any, **kwargs: Any) -> None:
+        """Send a message.
 
-    await notify_service.async_send_message(**kwargs)
+        kwargs can contain ATTR_TITLE to specify a title.
+        """
+        await self.hass.async_add_job(partial(self.send_message, message, **kwargs))  # type: ignore
+
+    async def _async_notify_message_service(self, service: ServiceCall) -> None:
+        """Handle sending notification message service calls."""
+        kwargs = {}
+        message = service.data[ATTR_MESSAGE]
+        title = service.data.get(ATTR_TITLE)
+
+        if title:
+            title.hass = self.hass
+            kwargs[ATTR_TITLE] = title.async_render()
+
+        if self._registered_targets.get(service.service) is not None:
+            kwargs[ATTR_TARGET] = [self._registered_targets[service.service]]
+        elif service.data.get(ATTR_TARGET) is not None:
+            kwargs[ATTR_TARGET] = service.data.get(ATTR_TARGET)
+
+        message.hass = self.hass
+        kwargs[ATTR_MESSAGE] = message.async_render()
+        kwargs[ATTR_DATA] = service.data.get(ATTR_DATA)
+
+        await self.async_send_message(**kwargs)
+
+    async def async_setup(
+        self,
+        hass: HomeAssistantType,
+        service_name: str,
+        target_service_name_prefix: str,
+    ) -> None:
+        """Store the data for the notify service."""
+        # pylint: disable=attribute-defined-outside-init
+        self.hass = hass
+        self._service_name = service_name
+        self._target_service_name_prefix = target_service_name_prefix
+        self._registered_targets: Dict = {}
+
+    async def async_register_services(self) -> None:
+        """Create or update the notify services."""
+        assert self.hass
```

Comment:
Codacy found an issue: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=6103615)